File: Niva_My_lang/basic.scala
// The file extension is scala for the sake of syntax highlighting

// 1 type declaration
type Person name: String age: Int

// 2 tagged union declaration
union User =
| LoggedIn name: String
| Guest

// 3 hello world
"Hello World!" echo

// 4 creating instance of type
person = Person name: "sas" age: 6

// 5 call function
person haveBirthday

// 7 create function for type
// 6 change instance field
// yes, assign can be used as expression here
Person haveBirthday = age <- age + 1


// 8 loop
1 to: 3 do: [ it echo ]
// 9 if
| true => "true" echo

// 10 switch

// smart casted inside branches 
User getName = | this
| Guest => "Guest" echo
| LoggedIn => this name 

// 11 list
list = {1 2 3}
list add: 4
// 1a 2a 3a 4a
mappedList = list map: [it toString + "a"]

// map 
map = #{1 "a" 2 "b"}
map key: 3 value: "c"
// "a"
a = map get: 1
| a != null => a echo

File: Nim/basic.nim
# 0 creating project
# just creat a file and run
# `nim r --hints=off file.nim`
# or
# `nimble init ProjectName`
# `nimble run`
# https://marketplace.visualstudio.com/items?itemName=nimsaem.nimvscode
{.experimental: "strictEffects".}
{.push raises:[] .}
# 1 type declaration
type Person = object
  name: string
  age: int

# 2 tagged union declaration
type 
  UserKind = enum LoggedIn, Guest
  User = object
    case kind: UserKind
    of LoggedIn:
      name: string
    of Guest:
      discard
  
# 3 Hello world
echo "Hello world!"

# 4 creating instance of type
var bob = Person(name: "Bob", age: 42)

# 7 create a function for type
proc haveBirthday(self: var Person) = # var means mutable argument
  # 6 change instance field
  self.age += 1
# "pure" functions are possible
func haveBirthdatyPure(self: Person): Person =
  Person(name: self.name, age: self.age + 1)

# 5 call function
bob.haveBirthday # same as haveBirthday(bob)

# call pure function
let immutableBob = Person(name: "Bob", age: 42)
let olderBob = immutableBob.haveBirthdatyPure

# 8 loop
for i in 1..10:
  echo i

# 9 if
proc sign(num: int): string =
  if num > 0: 
    "foo"
  elif num < 0:
    "bar"
  else:
    "42"

# 10 Switch
proc getName(x: User): string = 
  case x.kind:
  of LoggedIn:
    x.name
  of Guest:
    "Guest user"


# 11 array
var arr = @[1, 2] # @ makes in dynamic
arr.add 3

import std/[sequtils]
let 
  mappedArr = arr.mapIt(it * it)
  filtered = mappedArr.filterIt(it mod 2 == 0)
  both = mappedArr.concat filtered

# 12 Map
import std/tables
proc hg() =
  
  var table = {1: "a", 2: "b"}.toTable
  try:
    table[3] = "c"
    # Error, if not catch(push raises:[])
    echo table[4]
  except KeyError:
    echo "error"
  
# 13 import from another file
import anotherFile
let sum = 1.add 2

